### PR TITLE
[rllib] Add test for multi-agent support and fix IMPALA multi-agent

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

IMPALA support for multiagent was broken since IMPALA has a requirement that batch sizes be of a certain length. However multi-agent envs can create variable-length batches.

Fix this by adding zero-padding as needed (similar to the RNN case).

### CODE REVIEW

File: python/ray/rllib/models/model.py [hunk 1]
```diff
@@ -72,26 +72,28 @@ def __init__(self,
...
-        # Validate the output shape
+        if options.get("free_log_std", False):
+            log_std = tf.get_variable(
+                name="log_std",
+                shape=[num_outputs],
+                initializer=tf.zeros_initializer)
+            self.outputs = tf.concat(
+                [self.outputs, 0.0 * self.outputs + log_std], 1)
+
+    def _validate_output_shape(self):
```
Comment (ericl) [inline python/ray/rllib/models/model.py]: Make validation lazy we break validating lstm models.

**Comment by ericl** [comment]
jenkins retest this please

**Comment by ericl** [comment]
Tests look good.

**Comment by ericl** [comment]
Ping.

File: python/ray/rllib/evaluation/tf_policy_graph.py [hunk 2]
```diff
@@ -78,6 +79,9 @@ def __init__(self,
                 [NUM_SEQUENCES]. Note that NUM_SEQUENCES << BATCH_SIZE. See
                 models/lstm.py for more information.
             max_seq_len (int): max sequence length for LSTM training.
+            batch_divisibility_req (int): pad all agent experiences batches to
+                multiples of this value. This only has an effect if not already
```
Comment (richardliaw) [inline python/ray/rllib/evaluation/tf_policy_graph.py]: `if not using a LSTM model`?

Comment (richardliaw) [reply python/ray/rllib/evaluation/tf_policy_graph.py]: Sentence is a bit vague

File: python/ray/rllib/models/lstm.py [hunk 3]
```diff
@@ -52,15 +52,22 @@ def add_time_dimension(padded_inputs, seq_lens):
...
     """Truncate and pad experiences into fixed-length sequences.
 
     Arguments:
         episode_ids (list): List of episode ids for each step.
+        agent_indices (list): List of agent ids for each step. Note that this
+            has to be combined with episode_ids for uniqueness.
         feature_columns (list): List of arrays containing features.
         state_columns (list): List of arrays containing LSTM state values.
         max_seq_len (int): Max length of sequences before truncation.
+        dynamic_max (bool): Whether to dynamically shrink the max seq len.
```
Comment (richardliaw) [inline python/ray/rllib/models/lstm.py]: This could use an example of what this does exactly

**Review by richardliaw — APPROVED** [review]

**Pull request merged**